### ISSUE TITLE
Use Kotlin as future programming language for Soot

### ISSUE THREAD

**Issue by ericbodden** [soot-oss/soot#1050]
We, the core development team, have been discussing the possibility to use Kotlin as a development language at least for some parts of future (major) releases of Soot. In this issue we would like to get feedback from the community about whether or not this change would be desired.

We see the following Pros/Cons:
**Pro Kotlin:**
* Nicer language than Java, more concise, better concurrency constructs
* Features that do not work well are actually deprecated and discarded over time

**Neutral:**
* Kotlin can be integrated well with Java, in both directions

**Cons:**
* Steeper learning curve for people who wish to understand and edit Soot's code
* Possibly some cases where performance impact of the code are harder to predict than in Java

It would be great if you all could tell us your preferences and feedback, also regarding the above points.

**Comment by Leetsong** [comment]
Actually, I would be very appreciate if you could add Kotlin to Soot. 

Kotlin, though not popular enough, has already be the official of Android. It follows, in my opinion, the de facto spec of a modern language (however, maybe there is no standard spec or definition for it) that more and more classical languages will follow (e,g., the closure, the pattern matching, the flexibility). Its charming, and would become another dagger for Soot.

Well, the **Pros**:
1. Due to the official support, more and more Android developers are using Kotlin as their first language instead of Java. And many projects are moving to it. I've seen a lot around me (no data though, Jetbrains must have their report). And I don' t think it is long when the academia goes for it.
2. The modern features. Soot, as the first-class sharp object of static analysis tool in Java world, has already be a learning project for plain developers like me. It would be nice for us to learn your techniques in dealing with modern features (maybe too naive for most of you? alright, not for me).

For your concerns (the cons you mentioned), I'd suggest separate the Kotlin project out of Soot to be standalone project called like Soot-Kotlin. As far as I know, to extend Soot to Kotlin, what you guys need to do is compiling Kotlin to Jimple or the others, in both directions, whereas doing little in dealing with the static analysis, as Kotlin is built on JVM (Correct me if I was wrong). Thus, this is only a front-end work, separate it out of Soot can also make Soot incapable of dealing with other languages built on Soot. So, why not?

In conclusion, I I would be very appreciate if you could add Kotlin to Soot as a standalone project.

**Comment by cbruegg** [comment]
@Leetsong Although I very much agree with your points, according to my understanding, this issue is actually about using Kotlin as a language to _develop_ Soot.

**Comment by Leetsong** [comment]
@cbruegg Sorry for that. I misunderstood.

**Comment by Leetsong** [comment]
@cbruegg I am expecting the extension of Soot to Kotlin for long. Sorry again for my mistake.

**Comment by ericbodden** [comment]
Yes it would need to integrate with Java. Scala would be the only other option I could think of.

**Comment by cbruegg** [comment]
> Possibly some cases where performance impact of the code are harder to predict than in Java

I'd like to address this important point. Is there any particular language feature you're concerned about when it comes to performance? [This](https://sites.google.com/a/athaydes.com/renato-athaydes/posts/kotlinshiddencosts-benchmarks) article shows some JMH benchmarks for several language features that were suspected to cause performance penalties. However, the results don't seem to be particularly concerning.

Additionally, JetBrains has published various [official benchmarks](https://rawcdn.githack.com/JetBrains/kotlin-benchmarks/156d6912c5e3752a55ca76de8638b2f047c7bd75/report/report_java.html) comparing Kotlin's performance to Java. The code of the benchmarks can be found [here](https://github.com/JetBrains/kotlin-benchmarks).

Lastly, some anecdotal evidence: I've been using Kotlin in many Android applications in production, and as with any embedded device, the hardware performance of most Android devices is quite limited and yet users expect features such as smooth animations. Fortunately, I didn't run into any performance difficulties related to Kotlin yet, even though I mostly focus on implementing features and fixing bugs instead of performance optimization.

**Comment by cbruegg** [comment]
> Regarding Kotlin and speed I can only see from the Kotlin generated Android code I have seen so far that it "enhances" the code with a large number of assertions. From my feeling more that I would ever have to write in plain Java. Therefore execution speed my slightly decrease by changing to Kotlin.

This is true, Kotlin does insert calls e.g. to verify that values coming from Java are not null when they are expected to be not null by code written in Kotlin, i.e. an argument type is `String` and not `String?`. However, this is needed to prevent bugs. Secondly, the JVM is very good at inlining such verifications and can even eliminate null-checks. It simply registers a segfault-handler that handles accesses to variables that are null and resumes the program with a `NullPointerException` in case that happens. Only if this happens very often the JVM will deoptimize the method and actually perform the check. For more on this, I recommend [this talk about the JVM JIT](https://www.youtube.com/watch?v=oH4_unx8eJQ). It nicely demonstrates that arguing about performance on the JVM is non-trivial and benchmarks are always necessary.

> I thought Soot is just the opposite - never ever break compatibility at any point.

There would need to be very strong reasons for the Kotlin designers to remove a feature in a way that would make it impossible for Kotlin users to maintain a stable ABI and API. Therefore I'd say this is very unlikely to happen.
